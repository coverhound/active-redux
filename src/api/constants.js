export const API_CONFIGURE = 'AR_API_CONFIGURE';
export const API_CLEAR = 'AR_API_CLEAR';
export const API_HYDRATE = 'API_HYDRATE';

export const API_WILL_READ = 'AR_API_WILL_READ';
export const API_READ_DONE = 'AR_API_READ_DONE';
export const API_READ_FAILED = 'AR_API_READ_FAILED';

export const API_WILL_CREATE = 'AR_API_WILL_CREATE';
export const API_CREATE_DONE = 'AR_API_CREATE_DONE';
export const API_CREATE_FAILED = 'AR_API_CREATE_FAILED';

export const API_WILL_UPDATE = 'AR_API_WILL_UPDATE';
export const API_UPDATE_DONE = 'AR_API_UPDATE_DONE';
export const API_UPDATE_FAILED = 'AR_API_UPDATE_FAILED';

export const API_WILL_DELETE = 'AR_API_WILL_DELETE';
export const API_DELETE_DONE = 'AR_API_DELETE_DONE';
export const API_DELETE_FAILED = 'AR_API_DELETE_FAILED';

export const API_WILL_INDEX = 'AR_API_WILL_INDEX';
export const API_INDEX_DONE = 'AR_API_INDEX_DONE';
export const API_INDEX_CLEAR = 'AR_API_INDEX_CLEAR';
