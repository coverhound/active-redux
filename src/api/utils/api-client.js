import axios from 'axios';
import createError from 'axios/lib/core/createError';

const jsonContentTypes = [
  'application/json',
  'application/vnd.api+json',
];

const validateContentType = (response) => {
  const contentType = response.headers['content-type'].split(';')[0];
  if (jsonContentTypes.includes(contentType)) return;

  throw createError(
    'Invalid Content-Type in response',
    response.config,
    null,
    response
  );
};

const mergeDefaults = (url, options) => ({
  ...options,
  url,
  headers: {
    Accept: 'application/vnd.api+json',
    'Content-Type': 'application/vnd.api+json',
    ...options.headers,
  },
});

export default (url, options = {}) => (
  axios(mergeDefaults(url, options)).then((response) => {
    if (response.status === 204) return response;
    validateContentType(response);
    return response.data;
  })
);
