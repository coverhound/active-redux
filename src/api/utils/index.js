import imm from 'object-path-immutable';
import Registry from '../../registry';
import '../../polyfill';

export apiClient from './api-client';

/**
 * @module
 * @private
 */

/**
 * @example
 * incrementProperty(state, 'isReading')
 * @param {Object} state Object on which the key is incremented
 * @param {string} key Key to increment
 */
export const incrementProperty = (state, key) => (
  imm.set(state, key, state[key] + 1)
);

/**
 * @example
 * decrementProperty(state, 'isReading')
 * @param {Object} state Object on which the key is decremented
 * @param {string} key Key to decrement
 */
export const decrementProperty = (state, key) => (
  imm.set(state, key, state[key] - 1)
);

/**
 * @param {Array|Object} data
 */
export const resourcesArray = (data = []) => (
  Array.isArray(data) ? data : [data]
);

/**
 * @example
 * const obj = { foo: bar: { baz: 'hi' } }
 * getProperty(obj, 'foo', 'bar', 'baz') // => 'hi'
 * getProperty(obj, 'a', 'b', 'c') // => undefined
 *
 * @param {Object} object Object to traverse
 * @param {String[]} keys Keys to look up
 */
export const getProperty = (object, ...keys) => (
  keys.reduce((acc, key) => {
    if (acc === undefined) return acc;
    return acc[key];
  }, object)
);

/**
 * Creates reverse relationships for data merged into the state
 * @param {Object} state
 * @param {Object} newState Immutable Object via object-path-immutable
 * @param {Object} data JSON-API data object
 * @param {Object} data.id
 * @param {Object} data.type
 * @param {Object} data.relationships
 */
export const createReverseRelationships = (state, newState, { id, type, relationships = {} }) => {
  Object.values(relationships).forEach(({ data }) => {
    if (!data) return;

    const children = resourcesArray(data);
    const childModel = Registry.get(children[0].type);
    const childData = { type, id };
    if (!childModel.relationships[type]) return;

    const { key, isArray } = childModel.relationships[type];

    children.forEach(({ type: relType, id: relId }) => {
      if (!getProperty(state, 'resources', relType, relId)) return;
      const objectPath = ['resources', relType, relId, 'relationships', key];
      const existingRelationships = resourcesArray(getProperty(state, ...objectPath));

      if (!isArray) {
        newState.set([...objectPath, 'data'], childData);
        return;
      }

      if (existingRelationships.find((relationship) => relationship.id === id)) return;

      newState.push([...objectPath, 'data'], childData);
    });
  });
};

/**
 * Merges all resources into the state
 * @param {Object} state
 * @param {Object} payload JSON-API payload
 * @param {Object} payload.data
 * @param {Object} payload.included
 */
export const mergeResources = (state, { data, included = [] }) => {
  const newState = imm(state);

  resourcesArray(data).concat(included).forEach((dataObj) => {
    const { id, type } = dataObj;
    newState.set(['resources', type, String(id)], dataObj);
    createReverseRelationships(state, newState, dataObj);
  });

  return newState.value();
};

/**
 * This function clears differently based on what's passed in
 * Object: { type, id } -> this resource is cleared
 * Array: Each of the resources is cleared
 * String: '[type]' -> These resources are cleared
 * undefined: -> All resources are cleared
 * @param {Object} state
 * @param {Object[]|Object|string|undefined} data
 */
export const clearResources = (state, data) => {
  switch (typeof data) {
    case 'object': {
      if (Array.isArray(data)) return data.reduce(clearResources, state);
      const { type, id } = data;
      return imm.del(state, ['resources', type, id]);
    }
    case 'string': return imm.set(state, ['resources', data], {});
    case 'undefined': return imm.set(state, 'resources', {});
    default: return state;
  }
};

/**
 * @example
 * const CUSTOM_REDUX_ACTION = 'CUSTOM_REDUX_ACTION';
 * export const customReduxAction = createAction(CUSTOM_REDUX_ACTION); // => [function]
 * customReduxAction(payload) // => { type, payload }
 * @param {string} type Action type
 */
export const createAction = (type) => (payload) => ({ type, payload });

/**
 * @typedef {Object} ReducerFactory
 * @property {Object} map Action to function mapping
 * @property {Object} initialState Initial reducer state
 *
 * @param {ReducerFactory} base
 * @param {...ReducerFactory} extensions
 */
export const createReducer = ({ map, initialState }, ...extensions) => {
  const extend = (...addedExtensions) => (
    createReducer({ map, initialState }, ...extensions, ...addedExtensions)
  );

  const reducerFactory = imm({ map, initialState }).value();
  extensions.forEach(({ map: m, initialState: i }) => {
    Object.assign(reducerFactory.map, m);
    Object.assign(reducerFactory.initialState, i);
  });

  const reducer = (state = reducerFactory.initialState, action) => {
    const reduce = reducerFactory.map[action.type];
    return reduce === undefined ? state : reduce(state, action);
  };

  reducer.extend = extend;

  return reducer;
};

/**
 * @param {Object} state
 * @param {Object} resource
 * @param {Array|Object} resource.data
 */
export const markPendingResources = (state, { data }) => (
  resourcesArray(data).reduce((acc, resource) => {
    if (!getProperty(state, 'resources', resource.type, resource.id)) return acc;
    return acc.set(['resources', resource.type, resource.id, 'isPending'], true);
  }, imm(state)).value()
);
