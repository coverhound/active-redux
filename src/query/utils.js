import qs from 'qs';
import { createSelectorCreator, defaultMemoize } from 'reselect';

const isEqual = (a, b) => (
  typeof a === 'string' ?
    a === b :
    JSON.stringify(a) === JSON.stringify(b)
);

const createDeepEqualSelector = createSelectorCreator(
  defaultMemoize,
  isEqual,
);

/**
 * @alias module:active-redux.createQuerySelector
 * @function
 * @example
 *
 * import { createQuerySelector } from 'active-redux';
 * import { Comment } from '@models';
 *
 * const Comment = ({ comment }) => <p>{comment.body}</p>;
 *
 * // needed in order to have a unique selector per instance of the component
 * const mapStateToProps = () => {
 *   const querySelector = createQuerySelector(Comment, 'find');
 *   return (state, props) => ({
 *     // Arguments passed into querySelector are passed to `Comment#find`
 *     // and turned into a selector
 *     comments: querySelector(props.commentId)(state)
 *   });
 * };
 *
 * const mapDispatchToProps = () => ({});
 *
 * export default connect(mapStateToProps, mapDispatchToProps)(Comment);
 */
export const createQuerySelector = (promise, selector) => createDeepEqualSelector(
  (...args) => args,
  (args) => {
    promise(...args);
    const out = selector(...args);
    out.promise = promise;
    return out;
  }
);

export const generateEndpoint = (endpoint, query) => {
  let queryString = query;
  if (typeof query !== 'string') {
    queryString = qs.stringify(query);
  }
  return `${endpoint}?${queryString}`;
};
