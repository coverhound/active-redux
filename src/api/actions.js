import * as ActionTypes from './constants';
import { createAction, apiClient } from './utils';
/**
 * @module active-redux/api
 */

const wrappedApiRequest = ({
  resource,
  options,
  endpoint,
  dispatch,
  success,
  failure,
}) => apiClient(endpoint, options).then((json) => {
  dispatch(success(json));
  return Promise.resolve(json);
}).catch((error) => {
  const err = error;
  err.resource = resource;

  dispatch(failure(err));
  return Promise.reject(err);
});

/**
 * Configure the API
 * @see {@link https://github.com/mzabriskie/axios#request-config}
 * @function
 * @param {Object} config - Axios configuration
 */
export const apiConfigure = createAction(ActionTypes.API_CONFIGURE);

/**
 * Clear an entity from the store
 * @function
 * @private
 * @example
 * import { api } from 'active-redux';
 *
 * state.api.people
 * // => Object<String: Person>
 *
 * dispatch(api.apiClear('people'))
 *
 * state.api.people
 * // => {}
 * @param {Object} data - JSON-API data
 */
export const apiClear = createAction(ActionTypes.API_CLEAR);

/**
 * Hydrate the store from JSON-API
 * @function
 * @example
 * import { api } from 'active-redux';
 *
 * state.api.people
 * // => {}
 *
 * const data = [{
 *   type: 'people',
 *   id: '5',
 *   attributes: {
 *     name: 'Joe',
 *     age: 35,
 *   }
 * }];
 *
 * dispatch(api.apiHydrate(data))
 *
 * state.api.people
 * // => Object<String: Person>
 * @param {Object} data - JSON-API data
 */
export const apiHydrate = createAction(ActionTypes.API_HYDRATE);

const apiWillCreate = createAction(ActionTypes.API_WILL_CREATE);
const apiCreateDone = createAction(ActionTypes.API_CREATE_DONE);
const apiCreateFailed = createAction(ActionTypes.API_CREATE_FAILED);

/**
 * Create a resource via API
 * @function
 * @example
 * import { api } from 'active-redux';
 * import Person from '../models/person';
 *
 * Person.find({ id: '5' });
 * // => Promise<null>
 *
 * const person = new Person({ attributes: { name: 'Joe' } });
 *
 * dispatch(api.apiCreate({ resource: person })).then((json) => {
 *   // do something with the response json
 * }).catch((error) => {
 *   // handle error
 * });
 *
 * Person.find({ id: '5' });
 * // => Promise<Person>
 * @param {Object} args
 * @param {define~Model} args.resource
 * @param {string} [args.endpoint]
 */
export const apiCreate = ({ resource, endpoint = resource.endpoint('create') }) => (dispatch, getState) => {
  dispatch(apiWillCreate(resource));
  return wrappedApiRequest({
    resource,
    options: {
      ...getState().api.apiConfig,
      method: 'POST',
      data: { data: resource }
    },
    endpoint,
    dispatch,
    success: apiCreateDone,
    failure: apiCreateFailed,
  });
};

const apiWillRead = createAction(ActionTypes.API_WILL_READ);
const apiReadDone = createAction(ActionTypes.API_READ_DONE);
const apiReadFailed = createAction(ActionTypes.API_READ_FAILED);

/**
 * Read a resource via API
 * @function
 * @example
 * import { api } from 'active-redux';
 * import Person from '../models/person';
 *
 * Person.all();
 * // => Promise<[]>
 *
 * dispatch(api.apiRead({ resource: Person })).then((json) => {
 *   // do something with the response json
 * }).catch((error) => {
 *   // handle error
 * });
 *
 * Person.all();
 * // => Promise<Array<Person>>
 * @param {Object} args
 * @param {Model} args.resource
 * @param {string} [args.endpoint]
 */
export const apiRead = ({ resource, endpoint = resource.endpoint('read') }) => (dispatch, getState) => {
  dispatch(apiWillRead(resource));
  return wrappedApiRequest({
    resource,
    options: {
      ...getState().api.apiConfig,
      method: 'GET',
      data: { data: resource }
    },
    endpoint,
    dispatch,
    success: apiReadDone,
    failure: apiReadFailed,
  });
};

const apiWillUpdate = createAction(ActionTypes.API_WILL_UPDATE);
const apiUpdateDone = createAction(ActionTypes.API_UPDATE_DONE);
const apiUpdateFailed = createAction(ActionTypes.API_UPDATE_FAILED);

/**
 * Update a resource via API
 * @function
 * @example
 * import { api } from 'active-redux';
 * import Person from '../models/person';
 *
 * Person.find({ id: '5' });
 * // => Promise<Person(id='5' name='Joe')>
 *
 * const person = new Person({ id: '5', attributes: { name: 'Jimmy' } });
 *
 * dispatch(api.apiUpdate({ resource: person })).then((json) => {
 *   // do something with the response json
 * }).catch((error) => {
 *   // handle error
 * });
 *
 * Person.find({ id: '5' });
 * // => Promise<Person(id='5' name='Jimmy')>
 * @param {Object} args
 * @param {Model} args.resource
 * @param {string} [args.endpoint]
 */
export const apiUpdate = ({ resource, endpoint = resource.endpoint('update') }) => (
  (dispatch, getState) => {
    dispatch(apiWillUpdate(resource));
    return wrappedApiRequest({
      resource,
      options: {
        ...getState().api.apiConfig,
        method: 'PATCH',
        data: { data: resource }
      },
      endpoint,
      dispatch,
      success: apiUpdateDone,
      failure: apiUpdateFailed,
    });
  }
);

const apiWillDelete = createAction(ActionTypes.API_WILL_DELETE);
const apiDeleteDone = createAction(ActionTypes.API_DELETE_DONE);
const apiDeleteFailed = createAction(ActionTypes.API_DELETE_FAILED);

/**
 * Delete a resource via API
 * @function
 * @example
 * import { api } from 'active-redux';
 * import Person from '../models/person';
 *
 * Person.find({ id: '5' });
 * // => Promise<Person(id='5')>
 *
 * const person = new Person({ id: '5' });
 *
 * dispatch(api.apiDelete({ resource: person })).then((json) => {
 *   // do something with the response json
 * }).catch((error) => {
 *   // handle error
 * });
 *
 * Person.find({ id: '5' });
 * // => Promise<null>
 * @param {Object} args
 * @param {Model} args.resource
 * @param {string} [args.endpoint]
 */
export const apiDelete = ({ resource, endpoint = resource.endpoint('delete') }) => (
  (dispatch, getState) => {
    const options = {
      ...getState().api.apiConfig,
      method: 'DELETE',
    };

    dispatch(apiWillDelete(resource));
    return apiClient(endpoint, options).then((json) => {
      dispatch(apiDeleteDone(resource));
      return Promise.resolve(json);
    }).catch((error) => {
      const err = error;
      err.resource = resource;

      dispatch(apiDeleteFailed(err));
      return Promise.reject(err);
    });
  }
);

const apiWillIndex = createAction(ActionTypes.API_WILL_INDEX);
const apiIndexDone = createAction(ActionTypes.API_INDEX_DONE);
export const apiIndexClear = createAction(ActionTypes.API_INDEX_CLEAR);

export const apiIndexAsync = ({ hash, promise }) => (dispatch, getState) => {
  const indexExists = () => getState().api.indices[hash];

  if (indexExists() && indexExists().isFetching) return promise;

  dispatch(apiWillIndex(hash));

  return promise.then((resources) => {
    if (!indexExists()) return resources;
    dispatch(apiIndexDone({ hash, resources }));
    return resources;
  });
};
export const apiIndexSync = ({ hash, resources }) => (dispatch) => {
  dispatch(apiWillIndex(hash));
  dispatch(apiIndexDone({ hash, resources }));
};
