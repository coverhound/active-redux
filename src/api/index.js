import { createReducer } from './utils';
import baseReducer from './reducer';

export * from './reducer';

/**
 * API Reducer
 * @function
 * @alias module:active-redux/api
 * @example
 * import { createStore, combineReducers, applyMiddleware } from 'redux';
 * import { api } from 'active-redux';
 *
 * const initialState = {};
 *
 * export default createStore(
 *   combineReducers({ api: api.reducer }),
 *   initialState,
 *   applyMiddleware(thunk),
 * );
 *
 * @example
 * import { createStore, combineReducers, applyMiddleware } from 'redux';
 * import { api } from 'active-redux';
 *
 * const initialState = {};
 *
 * // You can add your own extensions to the reducer
 * const customExtensions = [{
 *   map: { AR_PAGINATE_MODELS: (state, payload) => {} },
 *   initialState: { pagination: {} },
 * }];
 *
 * export default createStore(
 *   combineReducers({ api: api.reducer.extend(...customExtension) }),
 *   initialState,
 *   applyMiddleware(thunk),
 * );
 * @param {Object} state
 * @param {Object} action
 */
export const reducer = createReducer(baseReducer);
