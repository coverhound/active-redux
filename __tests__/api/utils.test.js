jest.mock('active-redux/registry', () => ({
  people: {
    relationships: {
      comments: { key: 'comments', isArray: true }
    }
  },
  comments: {
    relationships: {
      people: { key: 'author', isArray: false }
    }
  },
  get(type) {
    return this[type];
  }
}));
import * as utils from 'active-redux/api/utils';
import jsonApiData from 'fixtures/json-api-body';
import imm from 'object-path-immutable';

const [person, ...comments] = jsonApiData.included;
const comment = comments[0];

describe('Utils', () => {
  describe('incrementProperty()', () => {
    test('increments the property', () => {
      const state = { isReading: 0 };
      const expected = { isReading: 1 };
      expect(utils.incrementProperty(state, 'isReading')).toEqual(expected);
    });

    test('doesn\'t mutate the original object', () => {
      const state = { isReading: 0 };
      utils.incrementProperty(state, 'isReading');
      expect(state.isReading).toEqual(0);
    });
  });

  describe('decrementProperty()', () => {
    test('decrements the property', () => {
      const state = { isReading: 1 };
      const expected = { isReading: 0 };
      expect(utils.decrementProperty(state, 'isReading')).toEqual(expected);
    });

    test('doesn\'t mutate the original object', () => {
      const state = { isReading: 0 };
      utils.decrementProperty(state, 'isReading');
      expect(state.isReading).toEqual(0);
    });
  });

  describe('createReverseRelationships()', () => {
    const state = {
      resources: {
        people: {
          9: {
            relationships: {
              comments: {
                data: [
                  { id: '12', type: 'comments' }
                ]
              }
            }
          }
        },
        comments: {
          5: {},
          12: {}
        },
      }
    };

    test('updates hasMany relationships', () => {
      const newState = imm(state);
      utils.createReverseRelationships(state, newState, comments[0]);
      const subject = newState.value().resources.people;

      const expectedData = [{ id: '12', type: 'comments' }, { id: '5', type: 'comments' }];
      expect(subject['9'].relationships.comments.data).toEqual(expectedData);
      expect(subject['9'].relationships).toMatchSnapshot();
    });

    test('updates hasOne relationships', () => {
      const newState = imm(state);
      utils.createReverseRelationships(state, newState, person);
      const subject = newState.value().resources.comments;

      const expectedData = { id: '9', type: 'people' };
      [subject['5'], subject['12']].forEach((model) => {
        expect(model.relationships.author.data).toEqual(expectedData);
      });
    });
  });

  describe('mergeResources()', () => {
    test('adds singular resources', () => {
      const resources = { data: person };
      expect(utils.mergeResources({}, resources)).toMatchSnapshot();
    });

    test('adds multiple resources', () => {
      const resources = { data: [person, comments[0]] };
      expect(utils.mergeResources({}, resources)).toMatchSnapshot();
    });

    test('adds included resources', () => {
      const resources = { data: [person], included: [comments[0]] };
      expect(utils.mergeResources({}, resources)).toMatchSnapshot();
    });

    test('doesn\'t mutate the original object', () => {
      const state = { resources: { [person.type]: 'foo' } };
      const resources = { data: person };
      utils.mergeResources(state, resources);
      expect(state).toEqual({ resources: { [person.type]: 'foo' } });
    });
  });

  describe('clearResources()', () => {
    const oldState = {
      resources: {
        [person.type]: { [person.id]: person },
        [comment.type]: { [comment.id]: comment },
      }
    };

    describe('given a resource', () => {
      test('it clears the resource', () => {
        const expected = {
          resources: {
            [person.type]: {},
            [comment.type]: { [comment.id]: comment }
          }
        };
        expect(utils.clearResources(oldState, person)).toEqual(expected);
      });
    });

    describe('given an array of resources', () => {
      test('it clears the resources', () => {
        const expected = {
          resources: {
            [person.type]: {},
            [comment.type]: {}
          }
        };
        expect(utils.clearResources(oldState, [person, comment])).toEqual(expected);
      });
    });

    describe('given a resource type', () => {
      test('it clears the resource type', () => {
        const expected = {
          resources: {
            [person.type]: {},
            [comment.type]: { [comment.id]: comment }
          }
        };
        expect(utils.clearResources(oldState, person.type)).toEqual(expected);
      });
    });

    describe('given no second argument', () => {
      test('it clears all resources', () => {
        const expected = { resources: {} };
        expect(utils.clearResources(oldState)).toEqual(expected);
      });
    });

    describe('doesn\'t mutate the original state', () => {
      const state = {
        resources: { [comment.type]: { [comment.id]: comment } }
      };
      const expected = JSON.parse(JSON.stringify(state));
      utils.clearResources(state);
      expect(state).toEqual(expected);
    });
  });

  describe('createAction()', () => {
    test('creates a Redux action', () => {
      const expected = { type: 'FOO', payload: 'BAR' };
      const action = utils.createAction(expected.type);
      expect(action(expected.payload)).toEqual(expected);
    });
  });

  describe('createReducer()', () => {
    const initialState = { foo: 'bar' };
    const type = 'FOO';

    describe('given an unmapped action', () => {
      test('returns the state', () => {
        const reducer = utils.createReducer({ map: {}, initialState: {} });
        expect(reducer(initialState, { type })).toEqual(initialState);
      });
    });

    describe('given a mapped action', () => {
      test('calls that action', () => {
        const reducer = utils.createReducer({
          map: {
            [type]: (state, { payload }) => ({
              foo: state.foo + payload
            })
          },
          initialState: {},
        });
        expect(reducer(initialState, { type, payload: 'baz' })).toEqual({
          foo: 'barbaz'
        });
      });
    });

    describe('when the state is undefined', () => {
      test('returns the initial state', () => {
        const reducer = utils.createReducer({ map: {}, initialState });
        expect(reducer(undefined, { type: 'FOO' })).toEqual(initialState);
      });
    });
  });

  describe('markPendingResources()', () => {
    const state = { resources: { people: {
      1: {},
      2: {},
      3: {},
    } } };

    test('marks a single resouurce', () => {
      const resource = { data: { type: 'people', id: 1 } };
      const expected = { resources: { people: {
        1: { isPending: true },
        2: {},
        3: {}
      } } };
      expect(utils.markPendingResources(state, resource)).toEqual(expected);
    });

    test('marks multiple resouurces', () => {
      const resources = { data: [
        { type: 'people', id: 1 },
        { type: 'people', id: 3 }
      ] };
      const expected = { resources: { people: {
        1: { isPending: true },
        2: {},
        3: { isPending: true }
      } } };
      expect(utils.markPendingResources(state, resources)).toEqual(expected);
    });

    test('doesn\'t mutate the original object', () => {
      const resource = { data: { type: 'people', id: 1 } };
      const expected = JSON.parse(JSON.stringify(state));
      utils.markPendingResources(state, resource);
      expect(state).toEqual(expected);
    });
  });
});
