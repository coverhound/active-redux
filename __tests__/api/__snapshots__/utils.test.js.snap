// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Utils createReverseRelationships() updates hasMany relationships 1`] = `
Object {
  "comments": Object {
    "data": Array [
      Object {
        "id": "12",
        "type": "comments",
      },
      Object {
        "id": "5",
        "type": "comments",
      },
    ],
  },
}
`;

exports[`Utils mergeResources() adds included resources 1`] = `
Object {
  "resources": Object {
    "comments": Object {
      "5": Object {
        "attributes": Object {
          "body": "First!",
        },
        "id": "5",
        "links": Object {
          "self": "http://example.com/comments/5",
        },
        "relationships": Object {
          "author": Object {
            "data": Object {
              "id": "9",
              "type": "people",
            },
          },
        },
        "type": "comments",
      },
    },
    "people": Object {
      "9": Object {
        "attributes": Object {
          "address": Object {
            "road": "123 Main St",
          },
          "first-name": "Dan",
          "last-name": "Gebhardt",
          "twitter": "dgeb",
        },
        "id": "9",
        "links": Object {
          "self": "http://example.com/people/9",
        },
        "relationships": Object {
          "comments": Object {
            "data": Array [
              Object {
                "id": "5",
                "type": "comments",
              },
              Object {
                "id": "12",
                "type": "comments",
              },
            ],
          },
        },
        "type": "people",
      },
    },
  },
}
`;

exports[`Utils mergeResources() adds multiple resources 1`] = `
Object {
  "resources": Object {
    "comments": Object {
      "5": Object {
        "attributes": Object {
          "body": "First!",
        },
        "id": "5",
        "links": Object {
          "self": "http://example.com/comments/5",
        },
        "relationships": Object {
          "author": Object {
            "data": Object {
              "id": "9",
              "type": "people",
            },
          },
        },
        "type": "comments",
      },
    },
    "people": Object {
      "9": Object {
        "attributes": Object {
          "address": Object {
            "road": "123 Main St",
          },
          "first-name": "Dan",
          "last-name": "Gebhardt",
          "twitter": "dgeb",
        },
        "id": "9",
        "links": Object {
          "self": "http://example.com/people/9",
        },
        "relationships": Object {
          "comments": Object {
            "data": Array [
              Object {
                "id": "5",
                "type": "comments",
              },
              Object {
                "id": "12",
                "type": "comments",
              },
            ],
          },
        },
        "type": "people",
      },
    },
  },
}
`;

exports[`Utils mergeResources() adds singular resources 1`] = `
Object {
  "resources": Object {
    "people": Object {
      "9": Object {
        "attributes": Object {
          "address": Object {
            "road": "123 Main St",
          },
          "first-name": "Dan",
          "last-name": "Gebhardt",
          "twitter": "dgeb",
        },
        "id": "9",
        "links": Object {
          "self": "http://example.com/people/9",
        },
        "relationships": Object {
          "comments": Object {
            "data": Array [
              Object {
                "id": "5",
                "type": "comments",
              },
              Object {
                "id": "12",
                "type": "comments",
              },
            ],
          },
        },
        "type": "people",
      },
    },
  },
}
`;
